packages/driver/cross-origin-testing.md

```diff
@@ -0,0 +1,137 @@
+- **origin**: The combination of the protocol, hostname, and port of a URL. For the purposes of Cypress, the subdomain is irrelevant. (e.g. http://example.com:3500)
+- **top**: The main window/frame of the browser
+- **primary origin**: The origin that top is on
+- **secondary origin**: Any origin that is not the primary origin
+- **primary driver**: The Cypress driver that run in **top** on the primary origin
+- **secondary driver**: Any Cypress driver that run in a **spec bridge**, interacting with a secondary origin
+
+## Frame architecture (outside of multi-domain)
+
+When testing a single origin, all 3 frames are loaded on that origin.
+
+Components (nesting shows hierarchical relationship):
+
+- **top (frame)**: The driver exists here.
+    - **spec frame**: The users’s spec code exists here.
+    - **AUT (frame)**: **A**pp **U**nder **T**est. The users’ app/site exists here.
+
+**top** communicates directly and synchronously with the **spec frame** and the **AUT frame**. The **spec frame** runs the spec, which uses the driver to run commands and interact with the **AUT**.
+
+In a single test (`it` + hooks), the **AUT** must remain on the same origin or a cross-origin error will occur, as **top** can no longer interact with the **AUT** in that circumstance. Different tests can visit different origins. In this case, we change **top** to the new origin, which also runs the **spec frame** on that origin. This navigation causes the spec to run again. We skip any already-run tests and resume.
+
+## Frame architecture (with multi-domain)
+
+Let’s say the primary origin is [domain1.com](http://domain1.com) and the secondary origin is domain2.com. The test has visited domain1.com and then issued a click that caused the **AUT** to navigate to domain2.com.
+
+Since the **AUT** is no longer on the same origin as **top**, they can no longer communicate synchronously. In order to facilitate cross-origin testing, we create another iframe we call the **spec bridge**. It exists as a sibling to the **AUT** (meaning they share the same parent frame, not to be confused with a DOM sibling). It contains a version of the driver tailored to cross-origin testing, with a different entry-point to the primary driver, but containing mostly the same code.
+
+The **spec bridge** is run on [domain2.com](http://domain2.com) to match the **AUT**. This and being a sibling allows the spec bridge to communicate directly with the **AUT**. The **spec bridge** communicates *asynchronously* with **top** via the [postMessage API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).
+
+The **spec bridge** remains in the DOM from the moment it’s created until the browser is refreshed or closed, the same as the the **spec frame** and **AUT**.
+
+Here’s what the components look like now (nesting shows hierarchical relationship):
+
+- **top (frame)**: domain1.om
+    - **spec frame**: domain1.com
+    - **AUT (frame)**: domain2.com
+    - **spec bridge**: domain2.com
```

Comment:
Could potentially use mermaid to demonstrate the frame relationships 🤷‍♂️ 

```mermaid
graph TD;
    top["top (frame): domain1.com"]-->specFrame["spec frame: domain1.com"];
    top-->aut["AUT (frame): domain2.com"];
    top-->specBridge["spec bridge: domain2.com"];
```

If not, you missed a `c` in the first domain
```suggestion
- **top (frame)**: domain1.com
    - **spec frame**: domain1.com
    - **AUT (frame)**: domain2.com
    - **spec bridge**: domain2.com
```

Reply:
Ooh good idea. Addressed in https://github.com/cypress-io/cypress/commit/0b0667591662fef71b485f6c7cf622e59406503d